File: genztelugu/translator.py
"""GenZ Telugu translation engine.

Translates GenZ Telugu syntax to valid Python and vice versa.
Uses token-based parsing to avoid replacing keywords inside strings and comments.
"""

import re
from genztelugu.keywords import KEYWORDS, REVERSE_KEYWORDS

_TOKEN_PATTERN = re.compile(
    r'(?P<triple_string>"""[\s\S]*?"""|\'\'\'[\s\S]*?\'\'\')'
    r'|(?P<string>"[^"\\]*(?:\\.[^"\\]*)*"|\'[^\'\\]*(?:\\.[^\'\\]*)*\')'
    r'|(?P<comment>#[^\r\n]*)'
    r'|(?P<identifier>[a-zA-Z_][a-zA-Z0-9_]*)'
    r'|(?P<other>[\s\S])'
)


def translate_to_python(code: str) -> str:
    """Translate GenZ Telugu code to valid Python.

    Args:
        code: Source code written in GenZ Telugu syntax.

    Returns:
        Valid Python code string.
    """

    def _replace(match: re.Match[str]) -> str:
        if match.group("identifier"):
            identifier = match.group("identifier")
            if identifier in KEYWORDS:
                return KEYWORDS[identifier]
        return match.group(0)

    return _TOKEN_PATTERN.sub(_replace, code)


def translate_to_genztelugu(code: str) -> str:
    """Translate Python code to GenZ Telugu syntax.

    Args:
        code: Valid Python source code.

    Returns:
        Code with Python keywords replaced by GenZ Telugu equivalents.
    """

    def _replace(match: re.Match[str]) -> str:
        if match.group("identifier"):
            identifier = match.group("identifier")
            if identifier in REVERSE_KEYWORDS:
                return REVERSE_KEYWORDS[identifier]
        return match.group(0)

    return _TOKEN_PATTERN.sub(_replace, code)

File: src/translator/translator.ts
export const KEYWORDS: Record<string, string> = {
  cheppu: 'print',
  okavela: 'if',
  lekapothe_okavela: 'elif',
  lekapothe: 'else',
  kosam: 'for',
  lo: 'in',
  paridhi: 'range',
  pani: 'def',
  tirigi: 'return',
  nijam: 'True',
  abaddam: 'False',
  yevaru: 'None',
  antha_varaku: 'while',
  vadhilesi: 'break',
  konasagu: 'continue',
  mariyu: 'and',
  leka: 'or',
  kaadhu: 'not',
  adugu: 'input',
  podavu: 'len',
  sankhya: 'int',
  padham: 'str',
  bhinna_sankhya: 'float',
  motha_sankhya: 'sum',
  vinu: 'import',
  gumpu: 'list',
  kurcho: 'pass',
  bokka: 'Exception',
};

export const REVERSE_KEYWORDS: Record<string, string> = Object.fromEntries(
  Object.entries(KEYWORDS).map(([gz, py]) => [py, gz])
);

/**
 * Translates GenZ Telugu code to standard Python.
 * Uses a token-based regex to avoid translating keywords inside strings and comments.
 */
export function translateToPython(code: string): string {
  const tokenRegex = /(?<tripleString>"""[\s\S]*?"""|'''[\s\S]*?''')|(?<string>"[^"\\]*(?:\\.[^"\\]*)*"|'[^'\\]*(?:\\.[^'\\]*)*')|(?<comment>#[^\r\n]*)|(?<identifier>[a-zA-Z_][a-zA-Z0-9_]*)|(?<other>[\s\S])/g;

  return code.replace(tokenRegex, (...args) => {
    const groups = args[args.length - 1] as Record<string, string>;
    
    if (groups.identifier) {
      const lowerId = groups.identifier;
      if (KEYWORDS[lowerId]) {
        return KEYWORDS[lowerId];
      }
    }
    return args[0];
  });
}

/**
 * Translates standard Python code back to GenZ Telugu.
 * Uses a token-based regex to avoid translating keywords inside strings and comments.
 */
export function translateToGenZTelugu(code: string): string {
  const tokenRegex = /(?<tripleString>"""[\s\S]*?"""|'''[\s\S]*?''')|(?<string>"[^"\\]*(?:\\.[^"\\]*)*"|'[^'\\]*(?:\\.[^'\\]*)*')|(?<comment>#[^\r\n]*)|(?<identifier>[a-zA-Z_][a-zA-Z0-9_]*)|(?<other>[\s\S])/g;

  return code.replace(tokenRegex, (...args) => {
    const groups = args[args.length - 1] as Record<string, string>;
    
    if (groups.identifier) {
      if (REVERSE_KEYWORDS[groups.identifier]) {
        return REVERSE_KEYWORDS[groups.identifier];
      }
    }
    return args[0];
  });
}
